components/service/glean/src/main/java/mozilla/components/service/glean/storages/EventsStorageEngine.kt

```diff
@@ -153,15 +202,145 @@ internal object EventsStorageEngine : StorageEngine {
+            getFile(storeName).forEachLine {
+                count += 1
+            }
+        } catch (e: FileNotFoundException) {
+            // pass
+        } catch (e: IOException) {
+            logger.warn("IOException while reading events from disk")
+        }
+
+        return count
+    }
+
+    /**
+     * Serializes an event to its JSON representation.
+     *
+     * @param category Event category
+     * @param name Event name
+     * @param msSinceStart The ms since the process started
+     * @param extra A `Map<String, String>` of extra values
+     * @return JSONArray representing the event
+     */
+    fun serializeEvent(
+        category: String,
+        name: String,
+        msSinceStart: Long,
+        extra: Map<String, String>?
+    ): JSONArray {
+        val eventData = JSONArray()
+        eventData.put(msSinceStart)
+        eventData.put(category)
+        eventData.put(name)
+        if (extra != null) {
+            eventData.put(JSONObject(extra))
+        }
+        return eventData
+    }
+
+    /**
+     * Unserializes an event in JSON into a RecordedEventData object.
+     *
+     * @param jsonContent The JSONArray containing the data for the event.
+     * @return event data object
+     */
+    fun unserializeEvent(jsonContent: JSONArray): RecordedEventData {
+        var extra: Map<String, String>? = jsonContent.optJSONObject(EXTRA_FIELD)?.let {
+            var extraValues: MutableMap<String, String> = mutableMapOf()
+            val names = it.names()
+            for (i in 0..(names.length() - 1)) {
+                extraValues[names.getString(i)] = it.getString(names.getString(i))
+            }
+            extraValues
+        }
+
+        return RecordedEventData(
+            jsonContent.getString(CATEGORY_FIELD),
+            jsonContent.getString(NAME_FIELD),
+            jsonContent.getLong(TIMESTAMP_FIELD),
+            extra
+        )
+    }
+
+    /**
+     * Writes an event to a single store on disk.  If the store
+     * has [Glean.configuration.maxEvents] elements, a ping is
+     * assembled immediately.
+     *
+     * @param storeName The name of the store to add the event to.
+     * @param eventContent A string of JSON as returned by [serializeEvent]
+     * @return Returns true if the event buffer filled up with
+     * [Glean.configuration.maxEvents] events
+     */
+    @Synchronized
+    fun writeEventToDisk(storeName: String, eventContent: String): Boolean {
+        var count = eventStoreCounts.get(storeName) ?: getEventCount(storeName)
+
+        try {
+            getFile(storeName).appendText("${eventContent}\n")
+        } catch (e: IOException) {
+            logger.warn("IOException while writing event to disk: $e")
+            return false
+        }
+
+        var newCount = count + 1
+        eventStoreCounts[storeName] = newCount
+
+        return (
+            newCount == Glean.configuration.maxEvents &&
+            Glean.assembleAndSerializePing(storeName)
```

Comment:
Can we use `Glean.sendPings` here instead of calling this and then manually enqueuing the upload worker? I'm not sure why we had it like that before. If there's any specific reason, would you kindly drop a comment in the code?